cloudevents/working-drafts/dds-format.md

```diff
@@ -0,0 +1,247 @@
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::URI_REF)"/>
+     <member name="ce_uri_reference" type="nonBasic" nonBasicTypeName="io::cloudevents::URI_Ref"/>
+   </case>
+ </union>
+
+ <struct name="DDS-CE-Attribute">
+   <member name="key" type="string" stringMaxLength="255"/>
+   <member name="value" type="nonBasic" nonBasicTypeName="io::cloudevents::AttributeValue"/>
+ </struct>
+
+ <typedef name="DDS-CE-Attributes" type="nonBasic" nonBasicTypeName="io::cloudevents::DDS-CE-Attribute" sequenceMaxLength="100"/>
+
+```
+In this model an attribute's name is used as the *key* and is associated
+with its *value* stored in the appropriately typed property.
+
+This approach allows attributes to be represented and transported
+with no loss of *type* information.
+
+## 3. Data
+
+The specification allows for data payloads of the following types to be explicitly represented:
+
+* binary
+* string
+
+```xml
+
+ <enum name="DataKind">
+   <enumerator name="BINARY"/>
+   <enumerator name="TEXT"/>
+ </enum>
+	  
+ <union name="Data" extensibility="final">
+   <discriminator type="nonBasic" nonBasicTypeName="io::cloudevents::DataKind"/>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::BINARY)"/>
+     <member name="binary_data" type="byte" sequenceMaxLength="100"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::TEXT)"/>
+     <member name="text_data" type="string" stringMaxLength="255"/>
+   </case>
+ </union>
+```
+
+* When the type of the data is binary the DataKind attribute MUST be set to BINARY.
+* When the type of the data is generic text the DataKind attribute MUST be set to TEXT.
+
+* It is noted here that when using DDS it is not necessary to specify the dataschema
+as the DDS protocol itself transmits the schema during the discovery handshake
+between event producers and consumers.
+
+## 4. Transport
+
+Transports that support content identification MUST use the following designation:
+
+```text
+   application/cloudevents+dds
+```
+
+## 5. Batch Format
+
+The DDS event format does not currently define a batch mode format.
+
+## 6. Examples
+
+The following code-snippet shows how messages in the DDS event format might
+be constructed and transmitted assuming the availability of some convenience
+methods.
+
+### 6.1 Binary message event data
+
+```javascript
+
+const emit = async () => {
+  
+  const type = "org.cncf.cloudevents.example";
+  const source = "urn:event:from:myapi/resource/123";
+  const time = new Date().toISOString();
+  
+  // cloudevent+dds / binary
+  const ce_dds_binary_obj = new CloudEvent({
+    specversion: "1.0",
+    id: "b46cf653-d48a-4b90-8dfa-355c01061364",
+    type,
+    source,
+    datacontenttype: "application/cloudevent+dds",
+    subject: "SQUARE",
+    time,
+    datakey:"Somekey",
+    datacontentencoding: "binary",
+    data: Buffer.from("Some text string" as string) // send the binary representation of a string
+    // [ext1Name]: ext1Value,
```

Comment:
We started digging a bit deeper into optional attributes and extensions and would like to defer support for these until a later revision of the protocol binding. 

The issue is that there is some overlap between optional attributes that can be specified at the DDS protocol level using DDS User Data, Topic Data and Group Data Quality of Service (QoS) settings. It is not exactly clear if these QoS (if they are set) should be propagated up to CE Optional Attributes or if CE Optional Attributes should be mapped down to DDS QoS. Similarly, for extension attributes there is some overlap between DDS QoS and CE Extension attributes. The clearest overlap is between the Expiry Time and DDS Lifespan - we would like to consider use cases to determine how best to map from one to the other. 

At first glance, it seemed like it would be possible to simply carry the OPTIONAL attributes through, but now we are not so sure that this will make sense for DDS->CloudEvents use cases.

Does this approach seem reasonable?

Reply:
Btw, the working code (no extension attributes) is here: https://github.com/protimabanerjee/sdk-javascript/tree/main/examples/dds-ex-simple